pkg/vm/engine/tae/containers/utils.go

```diff
@@ -79,296 +75,145 @@ func CloneWithBuffer(src Vector, buffer *bytes.Buffer, allocator ...MemAllocator
-			bs.Data = make([]byte, v.Length()*8)
-			logutil.Warn("[Moengine]", common.OperationField("MOToVector"),
-				common.OperandField("Col length is 0"))
-		} else {
-			bs.Data = types.EncodeFixedSlice(v.Col.([]types.Datetime), 8)
-		}
+		bs = movecToBytes[types.Datetime](v)
 	case types.T_timestamp:
-		if v.Col == nil || len(v.Col.([]types.Timestamp)) == 0 {
-			bs.Data = make([]byte, v.Length()*8)
-			logutil.Warn("[Moengine]", common.OperationField("MOToVector"),
-				common.OperandField("Col length is 0"))
-		} else {
-			bs.Data = types.EncodeFixedSlice(v.Col.([]types.Timestamp), 8)
-		}
+		bs = movecToBytes[types.Timestamp](v)
 	case types.T_decimal64:
-		if v.Col == nil || len(v.Col.([]types.Decimal64)) == 0 {
-			bs.Data = make([]byte, v.Length()*8)
-			logutil.Warn("[Moengine]", common.OperationField("MOToVector"),
-				common.OperandField("Col length is 0"))
-		} else {
-			bs.Data = types.EncodeFixedSlice(v.Col.([]types.Decimal64), 8)
-		}
+		bs = movecToBytes[types.Decimal64](v)
 	case types.T_decimal128:
-		if v.Col == nil || len(v.Col.([]types.Decimal128)) == 0 {
-			bs.Data = make([]byte, v.Length()*16)
-			logutil.Warn("[Moengine]", common.OperationField("MOToVector"),
-				common.OperandField("Col length is 0"))
-		} else {
-			bs.Data = types.EncodeFixedSlice(v.Col.([]types.Decimal128), 16)
-		}
+		bs = movecToBytes[types.Decimal128](v)
 	case types.T_uuid:
-		if v.Col == nil || len(v.Col.([]types.Uuid)) == 0 {
-			bs.Data = make([]byte, v.Length()*16)
-			logutil.Warn("[Moengine]", common.OperationField("MOToVector"),
-				common.OperandField("Col length is 0"))
-		} else {
-			bs.Data = types.EncodeFixedSlice(v.Col.([]types.Uuid), 16)
-		}
+		bs = movecToBytes[types.Uuid](v)
 	case types.T_TS:
-		if v.Col == nil || len(v.Col.([]types.TS)) == 0 {
-			bs.Data = make([]byte, v.Length()*types.TxnTsSize)
-			logutil.Warn("[Moengine]", common.OperationField("MOToVector"),
-				common.OperandField("Col length is 0"))
-		} else {
-			bs.Data = types.EncodeFixedSlice(v.Col.([]types.TS), types.TxnTsSize)
-		}
+		bs = movecToBytes[types.TS](v)
 	case types.T_Rowid:
-		if v.Col == nil || len(v.Col.([]types.Rowid)) == 0 {
-			bs.Data = make([]byte, v.Length()*types.RowidSize)
-			logutil.Warn("[Moengine]", common.OperationField("MOToVector"),
-				common.OperandField("Col length is 0"))
-		} else {
-			bs.Data = types.EncodeFixedSlice(v.Col.([]types.Rowid), types.RowidSize)
-		}
+		bs = movecToBytes[types.Rowid](v)
 	case types.T_char, types.T_varchar, types.T_json, types.T_blob:
-		if v.Col == nil {
-			bs.Data = make([]byte, 0)
-		} else {
-			vbs := movec.GetBytesVectorValues(v)
-			for _, v := range vbs {
-				bs.Append(v)
-			}
+		bs = NewBytes()
+		if v.Col != nil {
+			bs.Header, bs.Storage = movec.MustVarlenaRawData(v)
 		}
 	default:
 		panic(any(moerr.NewInternalError("%s not supported", v.Typ.String())))
 	}
+	var np *roaring64.Bitmap
 	if v.Nsp.Np != nil {
-		np := &roaring64.Bitmap{}
+		np = roaring64.New()
```

Comment:
Get rid of roaring?   Maybe next, post 0.6.   When we totally refactor vector.

Reply:
Do it post 0.6